lib/rosette/integrations/smartling-integration/smartling_puller.rb

```diff
@@ -1,135 +1,217 @@
 
-        attr_reader :rosette_config, :smartling_apis, :completed_files_map, :locale
+        DEFAULT_THREAD_POOL_SIZE = 15
+
+        attr_reader :rosette_config
+        attr_reader :repo_config, :serializer_id, :extractor_id
+        attr_reader :thread_pool_size, :logger
 
-        def initialize(rosette_config, smartling_apis, locale)
+        def initialize(rosette_config)
           @rosette_config = rosette_config
-          @smartling_apis = smartling_apis
-          @locale = locale
-          @completed_files_map = Hash.new { |h, key| h[key] = [] }
+          @thread_pool_size = DEFAULT_THREAD_POOL_SIZE
+          @logger = Rosette.logger
         end
 
-        def pull(repo_name, extractor_id)
-          each_file_for_repo(repo_name) do |file|
-            next unless file.repo_name == repo_name
-            next unless file_has_changed?(repo_name, file)
+        def set_repo_config(repo_config)
+          @repo_config = repo_config
+          self
+        end
 
-            potential_files = rosette_config.datastore.file_list_for_repo(repo_name)
+        def set_serializer_id(serializer_id)
+          @serializer_id = serializer_id
+          self
+        end
 
-            snapshot = Rosette::Core::Commands::RepoSnapshotCommand.new(rosette_config)
-              .set_repo_name(repo_name)
-              .set_commit_id(file.commit_id)
-              .set_paths(potential_files)
-              .execute
+        def set_extractor_id(extractor_id)
+          @extractor_id = extractor_id
+          self
+        end
 
-            snapshot_commit_ids = snapshot.values.uniq.compact
+        def set_thread_pool_size(size)
+          @thread_pool_size = size
+          self
+        end
 
-            rosette_config.datastore.add_or_update_commit_log_locale(
-              file.commit_id, locale, file.translated_count
-            )
+        def set_logger(logger)
+          @logger = logger
+          self
+        end
+
+        def pull
+          if thread_pool_size > 1
+            pull_asynchronously
+          else
+            pull_synchronously
+          end
+        end
+
+        private
 
-            # Update the phrase count to avoid discrepancies between this and
-            # the initial number reported when the file is pushed (uploaded).
-            # For some reason these numbers can differ, meaning the file is
-            # either over-translated (more translations than phrases) or
-            # under-translated (less translations than phrases), even though
-            # the Smartling UI reports 100% for all locales.
-            rosette_config.datastore.add_or_update_commit_log(
-              repo_name, file.commit_id, nil, Rosette::DataStores::PhraseStatus::PENDING,
-              file.phrase_count
+        def pull_synchronously
+          head_refs = repo_config.repo.all_head_refs
+          head_refs.each_with_index do |head_ref, idx|
+            pull_head(head_ref)
+            logger.info(
```

Comment:
consider logging in json, its the new hotness (plus it is easy to group and search logs)